lib/scnnr/connection.rb

```diff
@@ -31,22 +36,40 @@ def send_json(data)
     def send_request
       block = block_given? ? Proc.new : nil
       request = build_request(&block)
-      run_request(request)
+      with_retries do
+        Net::HTTP.start(@uri.host, @uri.port, use_ssl: use_ssl?) do |http|
+          @logger&.info("Started #{@method.upcase} #{@uri}")
+          http.request(request)
+        end
+      end
     end
 
     private
 
-    def run_request(request)
-      Net::HTTP.start(@uri.host, @uri.port, use_ssl: use_ssl?) do |http|
-        @logger&.info("Started #{@method.upcase} #{@uri}")
-        http.request(request)
+    def with_retries
+      yield
+    rescue Errno::ECONNRESET => e
+      retry_count ||= 0
+
+      if retry_count < @retry_limit
+        retry_count += 1
+        @logger&.info("Retrying to connect: #{@uri}, attempt: #{retry_count}")
+
+        sleep @sleep_on_retry
+        retry
       end
+
+      raise RequestFailed, "Failed to connect: #{@uri.host} (from: #{error_location}, original: #{e.inspect})"
```

Comment:
Additionally, I noticed, if we could read the error message, the message after `from:` would look like `Scnnr::Connection#rescue in with_retries`. Could we improve it?

Reply:
Hmm, I'll change it to `from: Scnnr::Connection#with_retries`.

Reply:
Ah I get it. I didn't think about that in this way. For sure it should be `#send_request` then. It's public and code that actually fails comes from it!

Reply:
I think`#send_request` is a good idea :)

Reply:
Sorry for cutting in. Telling where the error occurs isn't the responsibility of backtraces? Why we should describe it in the message?
If someone can understand the place where the error occurs from the message, s/he can read the backtrace as well.

Reply:
Ah, good point! I agree, if there is a backtrace, there is no point to describe the method in the message...

@RobertMaciejewski I "unresolved" the comment, could you take a look again?

Reply:
@hiroara 
Valid point. I think by adding that it will be easier to understand where the error comes from. It saves going thru backtrace. Let me ask you this: what do you think should be in the error details?

Reply:
Good question 😃 
In my opinion, we should describe the summary of the actual cause.
This can contain why, how to fix it, related information which is useful to debug (e.g. an actual string of invalid URL), and so on.

Ideally, users should be able to fix the problem without reading source code. The error message can be used here.
About the place where the error occurs, this may be used in the next step to debug the error.

Reply:
@hiroara @iusztin 
As discussed off Github I've added more user friendly error message [here](https://github.com/NEWROPE/scnnr-ruby/pull/23/commits/b15c773ef36c43c1d8176ac2bd05d8a3e603e628). Take a look!